### PR TITLE
fix #269253: Opening bass clef disappears when mid-measure treble clef is applied

### PR DESCRIPTION
**Pull request by mattmcclinch**
See [269253: Opening bass clef disappears when mid-measure treble clef is applied](https://musescore.org/en/node/269253).

Header clefs are often omitted from the MSCX because they can often be generated. But, for example, if a header clef in the first measure has been changed from the default clef for that staff, then it will appear in the MSCX. The problem is that there is not enough information in the MSCX to distinguish between a header clef and a mid-measure clef on the first tick of the measure, especially if the clef is the first element that Measure::read() encounters. But it should be safe to assume that the clef is not a header clef if it is not the first element encountered. Such clefs will be loaded correctly now.

For the sake of compatibility, a clef that is the first element that Measure::read() encounters will continue to be treated as a header clef, unless it has been saved with additional information to tell Measure::read() that it is not a header clef. Because of this, certain existing scores may contain clefs that are still loaded incorrectly. These clefs must only be corrected once, and saved. Afterwards, all clefs in the score will be correct when the score is loaded.

I did not originally intend to replace quite so many lines of code. Measure::read() and Score::addSystemHeader() each seem to try to do the right thing regarding mid-measure clefs on the first tick of the measure, but neither function gets it quite right in certain cases, and the code has become quite hard to follow. I believe that I have made the code much simpler and more readable, whereas a more conservative approach would have been more confusing and prone to error.

The master branch introduces a separate HeaderClef segment type, and inserts Clef segments later in the measure. As a result, Measure::read() and Measure::addSystemHeader() are greatly simplified in the master branch where clefs are concerned. But in the master branch it is currently not possible to create a mid-measure clef on the first tick of the measure. This is presumably an unintended result of recent changes to the inner workings of the clef insertion process.

### CODE REVIEW

**Comment by Jojo-Schmitz** [comment]
Why rebase? There doesn't seem to be any conflict. Just restarting the build should work I think?

**Comment by mattmcclinch** [comment]
Some tests are failing simply because Clef::write() is writing additional information in certain cases. I suppose I will have to modify the test files.

**Comment by anatoly-os** [comment]
@mattmcclinch could you please fix the tests?

**Comment by Jojo-Schmitz** [comment]
@mattmcclinch: without fixed tests it won't get merged into 2.3, I'm afraid

**Comment by mattmcclinch** [comment]
It's failing tst_compat for a different reason. I haven't been able to fix that one yet.

**Comment by anatoly-os** [comment]
I don't think we can merge it to 2.3 anyway.

**Comment by RobFog** [comment]
This is "work in progress" and should be marked as such, right?

**Comment by mattmcclinch** [comment]
Actually, since this will not make it into 2.x, I am going to go ahead and close it.
